sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -972,11 +979,86 @@ public void populateDisplayData(DisplayData.Builder builder) {
 
+    private boolean hasStatementAndSetter() {
+      return inner.getStatement() != null && inner.getPreparedStatementSetter() != null;
+    }
+
     @Override
     public PDone expand(PCollection<T> input) {
+      // fixme: validate invalid table input
+      if (input.hasSchema() && !hasStatementAndSetter()) {
+        checkArgument(inner.getTable() != null, "table cannot be null if statement is not provided");
+        Schema schema = input.getSchema();
+        List<Schema.Field> fields = getFilteredFields(schema);
+        inner = inner.withStatement(JdbcUtil.generateStatement(inner.getTable(), fields));
+        inner = inner.withPreparedStatementSetter(new AutoGeneratedPreparedStatementSetter(fields, input.getToRowFunction()));
+      }
+
       inner.expand(input);
       return PDone.in(input.getPipeline());
     }
+
+    private List<Schema.Field> getFilteredFields(Schema schema) {
+      List<Schema.Field> fields = schema.getFields();
+      Connection connection;
+      Schema tableSchema;
+      try {
+        connection = inner.getDataSourceProviderFn().apply(null).getConnection();
+        // Query table to get actual columns
+        ResultSet rs = connection.createStatement()
+                .executeQuery(String.format("SELECT * FROM %s", inner.getTable()));
+        tableSchema = SchemaUtil.toBeamSchema(rs.getMetaData());
+      } catch (Exception e) {
+        throw new RuntimeException("Error while determining columns from table: " + inner.getTable(), e);
+      }
+      fields = fields.stream().filter((field) ->
+              tableSchema.getFields().stream().anyMatch((ele) ->
+                      SchemaUtil.compareSchemaField(field, ele)))
+              .collect(Collectors.toList());
+      return fields;
+    }
+
+    private class AutoGeneratedPreparedStatementSetter implements PreparedStatementSetter<T> {
+
+      private List<Schema.Field> fields;
+      private SerializableFunction<T, Row> toRowFn;
+      private List<PreparedStatementSetCaller> preparedStatementFieldSetterList = new ArrayList<>();
+
+      AutoGeneratedPreparedStatementSetter(List<Schema.Field> fields, SerializableFunction<T, Row> toRowFn) {
+        this.fields = fields;
+        this.toRowFn = toRowFn;
+        populatePreparedStatementFieldSetter();
+      }
+
+      private void populatePreparedStatementFieldSetter() {
+        IntStream.range(0, fields.size()).forEach((index) -> {
+          Schema.FieldType fieldType = fields.get(index).getType();
+          preparedStatementFieldSetterList.add(JdbcUtil.getPreparedStatementSetCaller(fieldType));
+        });
+      }
+
+      @Override
+      public void setParameters(T element, PreparedStatement preparedStatement) throws Exception {
+        Row row = (element instanceof Row) ? (Row) element : toRowFn.apply(element);
+        IntStream.range(0, fields.size()).forEach((index) -> {
+          try {
+            preparedStatementFieldSetterList.get(index).set(row, preparedStatement, index);
```

Comment:
I think there's a logic bug here. Given that the list of schema fields is derived from filtering the original schema, there's no guarantee that the index of the field will match the index of the corresponding field in the row. You'll need to access the row by field name or store the actual field indices somehow,

Reply:
Done